{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Database versions: {'data': [{'ramp_db_version': '2.5.2', 'db_mod_date': '2024-03-27 08:19:31.246126', 'status': 'current', 'data_source_id': 'hmdb', 'data_source_name': 'HMDB', 'data_source_url': 'https://hmdb.ca/', 'data_source_version': 'v5.0 (2021-11-17)'}, {'ramp_db_version': '2.5.2', 'db_mod_date': '2024-03-27 08:19:31.246126', 'status': 'current', 'data_source_id': 'reactome', 'data_source_name': 'Reactome', 'data_source_url': 'https://reactome.org/', 'data_source_version': 'v87 (Dec 2023)'}, {'ramp_db_version': '2.5.2', 'db_mod_date': '2024-03-27 08:19:31.246126', 'status': 'current', 'data_source_id': 'wiki', 'data_source_name': 'WikiPathways', 'data_source_url': 'https://www.wikipathways.org/index.php/WikiPathways', 'data_source_version': 'v20240210 (2024-02-10)'}, {'ramp_db_version': '2.5.2', 'db_mod_date': '2024-03-27 08:19:31.246126', 'status': 'current', 'data_source_id': 'kegg', 'data_source_name': 'KEGG', 'data_source_url': 'https://www.genome.jp/kegg/', 'data_source_version': 'from HMDB (v5.0) (2021-11-17) '}, {'ramp_db_version': '2.5.2', 'db_mod_date': '2024-03-27 08:19:31.246126', 'status': 'current', 'data_source_id': 'chebi', 'data_source_name': 'ChEBI', 'data_source_url': 'https://www.ebi.ac.uk/chebi/', 'data_source_version': 'Release 231 (2024-03-01)'}, {'ramp_db_version': '2.5.2', 'db_mod_date': '2024-03-27 08:19:31.246126', 'status': 'current', 'data_source_id': 'lipidmaps', 'data_source_name': 'Lipid Maps', 'data_source_url': 'https://www.lipidmaps.org/', 'data_source_version': 'Release 2024-03-06'}, {'ramp_db_version': '2.5.2', 'db_mod_date': '2024-03-27 08:19:31.246126', 'status': 'current', 'data_source_id': 'rhea', 'data_source_name': 'Rhea', 'data_source_url': 'https://www.rhea-db.org/', 'data_source_version': 'Release 131 (2024-01-)'}], 'function_call': 'RaMP::getCurrentRaMPSourceDBVersions()'}\n",
      "Valid ID types: {'data': [{'analyteType': 'Metabolites', 'idTypes': 'hmdb, chebi, chemspider, kegg, pubchem, CAS, wikidata, LIPIDMAPS, lipidbank, swisslipids, plantfa, kegg_glycan, rhea-comp, polymer'}, {'analyteType': 'Genes/Proteins', 'idTypes': 'ensembl, gene_symbol, uniprot, entrez, hmdb, wikidata, EN, ncbiprotein, brenda, chebi'}], 'function_call': 'RaMP::getPrefixesFromAnalytes(\"metabolite\"); RaMP::getPrefixesFromAnalytes(\"gene\")'}\n",
      "\n",
      "=== Metabolite Analysis ===\n",
      "\n",
      "Chemical properties:\n",
      "{'data': [{'chem_data_source': 'hmdb', 'chem_source_id': 'hmdb:HMDB0000001', 'iso_smiles': '[H]OC(=O)[C@@]([H])(N([H])[H])C([H])([H])C1=C([H])N(C([H])=N1)C([H])([H])[H]', 'inchi_key_prefix': 'BRMWTNUJHUMWMS', 'inchi_key': 'BRMWTNUJHUMWMS-LURJTMIESA-N', 'inchi': 'InChI=1S/C7H11N3O2/c1-10-3-5(9-4-10)2-6(8)7(11)12/h3-4,6H,2,8H2,1H3,(H,11,12)/t6-/m0/s1', 'mw': 169.1811, 'monoisotop_mass': 169.0851, 'common_name': '1-Methylhistidine', 'mol_formula': 'C7H11N3O2'}, {'chem_data_source': 'hmdb', 'chem_source_id': 'hmdb:HMDB0000002', 'iso_smiles': 'NCCCN', 'inchi_key_prefix': 'XFNJVJPLKCPIBV', 'inchi_key': 'XFNJVJPLKCPIBV-UHFFFAOYSA-N', 'inchi': 'InChI=1S/C3H10N2/c4-2-1-3-5/h1-5H2', 'mw': 74.1249, 'monoisotop_mass': 74.0844, 'common_name': '1,3-Diaminopropane', 'mol_formula': 'C3H10N2'}], 'function_call': ['RaMP::getChemicalProperties(c(\"hmdb:HMDB0000001\", \"hmdb:HMDB0000002\"))\\n'], 'numFoundIds': [2]}\n",
      "\n",
      "Chemical classes:\n",
      "{'data': [{'sourceId': 'hmdb:HMDB0000001', 'common_names': '1-Methylhistidine', 'class_level_name': 'ClassyFire_class', 'class_name': 'Carboxylic acids and derivatives', 'source': 'hmdb', 'directIdClassHits': 2}, {'sourceId': 'hmdb:HMDB0000001', 'common_names': '1-Methylhistidine', 'class_level_name': 'ClassyFire_sub_class', 'class_name': 'Amino acids, peptides, and analogues', 'source': 'hmdb', 'directIdClassHits': 2}, {'sourceId': 'hmdb:HMDB0000001', 'common_names': '1-Methylhistidine', 'class_level_name': 'ClassyFire_super_class', 'class_name': 'Organic acids and derivatives', 'source': 'hmdb', 'directIdClassHits': 2}, {'sourceId': 'hmdb:HMDB0000002', 'common_names': '1,3-Diaminopropane', 'class_level_name': 'ClassyFire_class', 'class_name': 'Organonitrogen compounds', 'source': 'hmdb', 'directIdClassHits': 1}, {'sourceId': 'hmdb:HMDB0000002', 'common_names': '1,3-Diaminopropane', 'class_level_name': 'ClassyFire_sub_class', 'class_name': 'Amines', 'source': 'hmdb', 'directIdClassHits': 1}, {'sourceId': 'hmdb:HMDB0000002', 'common_names': '1,3-Diaminopropane', 'class_level_name': 'ClassyFire_super_class', 'class_name': 'Organic nitrogen compounds', 'source': 'hmdb', 'directIdClassHits': 1}], 'function_call': ['RaMP::chemicalClassSurvey(c(\"hmdb:HMDB0000001\", \"hmdb:HMDB0000002\"))\\n'], 'numFoundIds': [2]}\n",
      "No pathway data found\n",
      "\n",
      "=== Protein Pathway Analysis ===\n",
      "\n",
      "Protein pathways:\n",
      "{'data': [{'pathwayName': '11p11.2 copy number variation syndrome', 'pathwaySource': 'wiki', 'pathwayId': 'WP5348', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': '16p11.2 proximal deletion syndrome', 'pathwaySource': 'wiki', 'pathwayId': 'WP4949', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': '8q11.23 (RB1CC1) copy number variation', 'pathwaySource': 'wiki', 'pathwayId': 'WP5412', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'A network map of Macrophage-stimulating protein (MSP) signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP5353', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Activation of BH3-only proteins', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-114452', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Activation of NMDA receptors and postsynaptic events', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-442755', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Activation of NOXA and translocation to mitochondria', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-111448', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Activation of PUMA and translocation to mitochondria', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-139915', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'ADORA2B mediated anti-inflammatory cytokines production', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9660821', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'AMP-activated protein kinase signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP1403', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Amyotrophic lateral sclerosis (ALS)', 'pathwaySource': 'wiki', 'pathwayId': 'WP2447', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Anchoring of the basal body to the plasma membrane', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5620912', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Androgen receptor network in prostate cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP2263', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Apoptosis', 'pathwaySource': 'wiki', 'pathwayId': 'WP254', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Apoptosis modulation and signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP1772', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Aquaporin-mediated transport', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-445717', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Association of TriC/CCT with target proteins during biosynthesis', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-390471', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'ATM signaling pathway', 'pathwaySource': 'wiki', 'pathwayId': 'WP2516', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'AURKA Activation by TPX2', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-8854518', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Autodegradation of the E3 ubiquitin ligase COP1', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-349425', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Autophagy in pancreatic ductal adenocarcinoma', 'pathwaySource': 'wiki', 'pathwayId': 'WP5331', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Bladder cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP2828', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Burn wound healing', 'pathwaySource': 'wiki', 'pathwayId': 'WP5055', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Ca-dependent events', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-111996', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Calmodulin induced events', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-111933', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Caloric restriction and aging', 'pathwaySource': 'wiki', 'pathwayId': 'WP4191', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'CaM pathway', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-111997', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Cannabinoid receptor signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP3869', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Cell cycle', 'pathwaySource': 'wiki', 'pathwayId': 'WP179', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Centrosome maturation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-380287', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Chaperonin-mediated protein folding', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-390466', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Chromosomal and microsatellite instability in colorectal cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP4216', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'CKAP4 signaling pathway map', 'pathwaySource': 'wiki', 'pathwayId': 'WP5322', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Copper homeostasis', 'pathwaySource': 'wiki', 'pathwayId': 'WP3286', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'CREB1 phosphorylation through the activation of Adenylate Cyclase', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-442720', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'DAG and IP3 signaling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-1489509', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'DAG and IP3 signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP2688', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'DARPP-32 events', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-180024', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Defective Intrinsic Pathway for Apoptosis', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9734009', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Diseases of programmed cell death', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9645723', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DNA damage response', 'pathwaySource': 'wiki', 'pathwayId': 'WP707', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DNA damage response (only ATM dependent)', 'pathwaySource': 'wiki', 'pathwayId': 'WP710', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DNA Damage/Telomere Stress Induced Senescence', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-2559586', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DNA Damage/Telomere Stress Induced Senescence', 'pathwaySource': 'wiki', 'pathwayId': 'WP3565', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DNA Double Strand Break Response', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5693606', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DNA Double Strand Break Response', 'pathwaySource': 'wiki', 'pathwayId': 'WP3543', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DNA IR-damage and cellular response via ATR', 'pathwaySource': 'wiki', 'pathwayId': 'WP4016', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DNA IR-double strand breaks and cellular response via ATM', 'pathwaySource': 'wiki', 'pathwayId': 'WP3959', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'DYRK1A involvement regarding cell proliferation in brain development', 'pathwaySource': 'wiki', 'pathwayId': 'WP5180', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Effect of progerin on genes involved in Hutchinson-Gilford progeria syndrome', 'pathwaySource': 'wiki', 'pathwayId': 'WP4320', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Endometrial cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP4155', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'ErbB signaling pathway', 'pathwaySource': 'wiki', 'pathwayId': 'WP673', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Factors involved in megakaryocyte development and platelet production', 'pathwaySource': 'wiki', 'pathwayId': 'WP1815', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Factors involved in megakaryocyte development and platelet production', 'pathwaySource': 'wiki', 'pathwayId': 'WP1815', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'FCGR3A-mediated IL10 synthesis', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9664323', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Ferroptosis', 'pathwaySource': 'wiki', 'pathwayId': 'WP4313', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Fluoropyrimidine activity', 'pathwaySource': 'wiki', 'pathwayId': 'WP1601', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Folate metabolism', 'pathwaySource': 'wiki', 'pathwayId': 'WP176', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Formation of Senescence-Associated Heterochromatin Foci (SAHF)', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-2559584', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'G protein signaling pathways', 'pathwaySource': 'wiki', 'pathwayId': 'WP35', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'G-protein mediated events', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-112040', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'G1 to S cell cycle control', 'pathwaySource': 'wiki', 'pathwayId': 'WP45', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'G1/S DNA Damage Checkpoints', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-69615', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'G2/M DNA damage checkpoint', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-69473', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Gastric cancer network 2', 'pathwaySource': 'wiki', 'pathwayId': 'WP2363', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Glioblastoma signaling pathways', 'pathwaySource': 'wiki', 'pathwayId': 'WP2261', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Glucagon signaling in metabolic regulation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-163359', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Glucagon-like Peptide-1 (GLP1) regulates insulin secretion', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-381676', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Glycolysis in senescence', 'pathwaySource': 'wiki', 'pathwayId': 'WP5049', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'GPER1 signaling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9634597', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Head and neck squamous cell carcinoma', 'pathwaySource': 'wiki', 'pathwayId': 'WP4674', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': \"Hedgehog 'off' state\", 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5610787', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': \"Hedgehog 'off' state\", 'pathwaySource': 'wiki', 'pathwayId': 'WP3316', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Hepatitis C and hepatocellular carcinoma', 'pathwaySource': 'wiki', 'pathwayId': 'WP3646', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Hereditary leiomyomatosis and renal cell carcinoma pathway', 'pathwaySource': 'wiki', 'pathwayId': 'WP4206', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Hippo signaling regulation pathways', 'pathwaySource': 'wiki', 'pathwayId': 'WP4540', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Hippo-Merlin signaling dysregulation', 'pathwaySource': 'wiki', 'pathwayId': 'WP4541', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Integrated cancer pathway', 'pathwaySource': 'wiki', 'pathwayId': 'WP1971', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Integration of energy metabolism', 'pathwaySource': 'wiki', 'pathwayId': 'WP1831', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Interleukin-1 signaling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9020702', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Interleukin-4 and Interleukin-13 signaling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6785807', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Interleukin-4 and Interleukin-13 signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP4066', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Intrinsic Pathway for Apoptosis', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-109606', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Intrinsic Pathway for Apoptosis', 'pathwaySource': 'wiki', 'pathwayId': 'WP1841', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Kynurenine pathway and links to cell senescence', 'pathwaySource': 'wiki', 'pathwayId': 'WP5044', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Lipid metabolism in senescent cells', 'pathwaySource': 'wiki', 'pathwayId': 'WP5149', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Lipid metabolism pathway', 'pathwaySource': 'wiki', 'pathwayId': 'WP3965', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'lncRNA-mediated mechanisms of therapeutic resistance', 'pathwaySource': 'wiki', 'pathwayId': 'WP3672', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Loss of Nlp from mitotic centrosomes', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-380259', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Loss of proteins required for interphase microtubule organization from the centrosome', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-380284', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Mammary gland development pathway - Involution (Stage 4 of 4)', 'pathwaySource': 'wiki', 'pathwayId': 'WP2815', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Maternal to zygotic transition (MZT)', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9816359', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Measles virus infection', 'pathwaySource': 'wiki', 'pathwayId': 'WP4630', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Melanoma', 'pathwaySource': 'wiki', 'pathwayId': 'WP4685', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Metabolic reprogramming in pancreatic cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP5220', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Metastatic brain tumor', 'pathwaySource': 'wiki', 'pathwayId': 'WP2249', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'MicroRNA network associated with chronic lymphocytic leukemia', 'pathwaySource': 'wiki', 'pathwayId': 'WP4399', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'miRNA regulation of DNA damage response', 'pathwaySource': 'wiki', 'pathwayId': 'WP1530', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'miRNA regulation of p53 pathway in prostate cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP3982', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'miRNA regulation of prostate cancer signaling pathways', 'pathwaySource': 'wiki', 'pathwayId': 'WP3981', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'miRNAs involved in DNA damage response', 'pathwaySource': 'wiki', 'pathwayId': 'WP1545', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'MyD88 cascade initiated on plasma membrane', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-975871', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'MyD88 dependent cascade initiated on endosome', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-975155', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'MyD88-independent TLR4 cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-166166', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'MyD88:MAL(TIRAP) cascade initiated on plasma membrane', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-166058', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'NAD metabolism in oncogene-induced senescence and mitochondrial dysfunction-associated senescence', 'pathwaySource': 'wiki', 'pathwayId': 'WP5046', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Netrin-UNC5B signaling pathway', 'pathwaySource': 'wiki', 'pathwayId': 'WP4747', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Non-small cell lung cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP4255', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Oncogene Induced Senescence', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-2559585', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Oncogene Induced Senescence', 'pathwaySource': 'wiki', 'pathwayId': 'WP3308', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Opioid signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP1978', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Opioid Signalling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-111885', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Ovarian tumor domain proteases', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5689896', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Oxidative Stress Induced Senescence', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-2559580', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Oxidative Stress Induced Senescence', 'pathwaySource': 'wiki', 'pathwayId': 'WP3404', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'p53-Dependent G1 DNA Damage Response', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-69563', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'p53-Dependent G1/S DNA damage checkpoint', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-69580', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Pancreatic adenocarcinoma pathway', 'pathwaySource': 'wiki', 'pathwayId': 'WP4263', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Pathways affected in adenoid cystic carcinoma', 'pathwaySource': 'wiki', 'pathwayId': 'WP3651', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Pentose phosphate pathway in senescent cells', 'pathwaySource': 'wiki', 'pathwayId': 'WP5043', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Photodynamic therapy-induced AP-1 survival signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP3611', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Photodynamic therapy-induced HIF-1 survival signaling', 'pathwaySource': 'wiki', 'pathwayId': 'WP3614', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'PI5P Regulates TP53 Acetylation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6811555', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'PKA activation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-163615', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'PKA activation in glucagon signalling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-164378', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'PKA-mediated phosphorylation of CREB', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-111931', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'PKR-mediated signaling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9833482', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'PLC beta mediated events', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-112043', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Post NMDA receptor activation events', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-438064', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Prader-Willi and Angelman syndrome', 'pathwaySource': 'wiki', 'pathwayId': 'WP3998', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Pre-NOTCH Expression and Processing', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-1912422', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Pre-NOTCH Expression and Processing', 'pathwaySource': 'wiki', 'pathwayId': 'WP2786', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Pre-NOTCH Transcription and Translation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-1912408', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Protein folding', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-391251', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Protein folding', 'pathwaySource': 'wiki', 'pathwayId': 'WP1892', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Pyroptosis', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5620971', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'RAC1/PAK1/p38/MMP2 pathway', 'pathwaySource': 'wiki', 'pathwayId': 'WP3303', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Recruitment and ATM-mediated phosphorylation of repair and signaling proteins at DNA double strand breaks', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5693565', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Recruitment of mitotic centrosome proteins and complexes', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-380270', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Recruitment of NuMA to mitotic centrosomes', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-380320', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Regulated Necrosis', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5218859', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of insulin secretion', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-422356', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Regulation of NF-kappa B signaling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9758274', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of PLK1 Activity at G2/M Transition', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-2565942', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Regulation of PTEN gene transcription', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-8943724', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Activity through Acetylation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6804758', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Activity through Acetylation', 'pathwaySource': 'wiki', 'pathwayId': 'WP3816', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Activity through Association with Co-factors', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6804759', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 activity through association with cofactors', 'pathwaySource': 'wiki', 'pathwayId': 'WP3798', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Activity through Methylation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6804760', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Activity through Methylation', 'pathwaySource': 'wiki', 'pathwayId': 'WP3823', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Activity through Phosphorylation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6804756', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Activity through Phosphorylation', 'pathwaySource': 'wiki', 'pathwayId': 'WP3838', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Degradation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6804757', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Expression and Degradation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6806003', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Regulation of TP53 Expression and Degradation', 'pathwaySource': 'wiki', 'pathwayId': 'WP3807', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Retinoblastoma gene in cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP2446', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'RUNX3 regulates CDKN1A transcription', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-8941855', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Signaling by ALK fusions and activated point mutants', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9725370', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Signaling by ALK in cancer', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9700206', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Small cell lung cancer', 'pathwaySource': 'wiki', 'pathwayId': 'WP4658', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Spinal cord injury', 'pathwaySource': 'wiki', 'pathwayId': 'WP2431', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Stabilization of p53', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-69541', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'SUMOylation of transcription factors', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-3232118', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'SUMOylation of transcription factors', 'pathwaySource': 'wiki', 'pathwayId': 'WP3824', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'T-cell activation SARS-CoV-2', 'pathwaySource': 'wiki', 'pathwayId': 'WP5098', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TAK1-dependent IKK and NF-kappa-B activation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-445989', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TCA cycle in senescence', 'pathwaySource': 'wiki', 'pathwayId': 'WP5050', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'The role of GTSE1 in G2/M progression after G2 checkpoint', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-8852276', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Thyroid hormones production and peripheral downstream signaling effects', 'pathwaySource': 'wiki', 'pathwayId': 'WP4746', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Toll Like Receptor 10 (TLR10) Cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-168142', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Toll Like Receptor 2 (TLR2) Cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-181438', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Toll Like Receptor 3 (TLR3) Cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-168164', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Toll Like Receptor 5 (TLR5) Cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-168176', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Toll Like Receptor 7/8 (TLR7/8) Cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-168181', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Toll Like Receptor 9 (TLR9) Cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-168138', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Toll Like Receptor TLR1:TLR2 Cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-168179', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Toll Like Receptor TLR6:TLR2 Cascade', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-168188', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 network', 'pathwaySource': 'wiki', 'pathwayId': 'WP1742', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Metabolic Genes', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5628897', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Metabolic Genes', 'pathwaySource': 'wiki', 'pathwayId': 'WP3349', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 regulates transcription of additional cell cycle genes whose exact role in the p53 pathway remain uncertain', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6804115', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of Caspase Activators and Caspases', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6803207', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of Cell Cycle Genes', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6791312', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 regulates transcription of cell cycle genes', 'pathwaySource': 'wiki', 'pathwayId': 'WP3804', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of Cell Death Genes', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-5633008', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of Cell Death Genes', 'pathwaySource': 'wiki', 'pathwayId': 'WP3802', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of Death Receptors and Ligands', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6803211', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of DNA Repair Genes', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6796648', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of DNA Repair Genes', 'pathwaySource': 'wiki', 'pathwayId': 'WP3808', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of Genes Involved in Cytochrome C Release', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6803204', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of Genes Involved in G1 Cell Cycle Arrest', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6804116', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 Regulates Transcription of Genes Involved in G2 Cell Cycle Arrest', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6804114', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TP53 regulates transcription of several additional cell death genes whose specific roles in p53-dependent apoptosis remain uncertain', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-6803205', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TRAF6 mediated induction of NFkB and MAP kinases upon TLR7/8 or 9 activation', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-975138', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Transcriptional regulation by RUNX3', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-8878159', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Transcriptional regulation by RUNX3', 'pathwaySource': 'wiki', 'pathwayId': 'WP4107', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Transcriptional Regulation by VENTX', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-8853884', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Transcriptional regulation by VENTX', 'pathwaySource': 'wiki', 'pathwayId': 'WP4995', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'TRIF(TICAM1)-mediated TLR4 signaling', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-937061', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Unfolded protein response', 'pathwaySource': 'wiki', 'pathwayId': 'WP4925', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Vasopressin regulates renal water homeostasis via Aquaporins', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-432040', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Vasopressin regulates renal water homeostasis via aquaporins', 'pathwaySource': 'wiki', 'pathwayId': 'WP2662', 'inputId': 'uniprot:P31323', 'commonName': 'PRKAR2B'}, {'pathwayName': 'Wnt signaling pathway and pluripotency', 'pathwaySource': 'wiki', 'pathwayId': 'WP399', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}, {'pathwayName': 'Zygotic genome activation (ZGA)', 'pathwaySource': 'reactome', 'pathwayId': 'R-HSA-9819196', 'inputId': 'uniprot:P04637', 'commonName': 'TP53'}], 'function_call': ['RaMP::getPathwayFromAnalyte(c(\"uniprot:P31323\", \"uniprot:P04637\"))\\n'], 'numFoundIds': [2]}\n",
      "No pathway data found\n",
      "\n",
      "=== Combined Analysis ===\n",
      "\n",
      "Common reaction analytes:\n",
      "{'data': [{'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000001', 'input_common_names': '1-Methylhistidine', 'rxn_partner_common_name': 'CNDP1', 'rxn_partner_ids': 'hmdb:HMDBP00473,gene_symbol:CNDP1,uniprot:Q96KN2,uniprot:Q14D40,uniprot:Q17S05,uniprot:Q2TBG0,uniprot:Q6UWK2,uniprot:Q9BT98'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000001', 'input_common_names': '1-Methylhistidine', 'rxn_partner_common_name': 'PRMT3', 'rxn_partner_ids': 'hmdb:HMDBP00474,gene_symbol:PRMT3,uniprot:O60678,entrez:10196,ensembl:ENSG00000185238,uniprot:E9PRB4,uniprot:E9PLF6,uniprot:A0A0A0MSN7,uniprot:B4DUC7'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000002', 'input_common_names': '1,3-Diaminopropane', 'rxn_partner_common_name': 'SMS', 'rxn_partner_ids': 'hmdb:HMDBP00217,gene_symbol:SMS,uniprot:P52788,entrez:6611,ensembl:ENSG00000102172,uniprot:H7C2R7,uniprot:A6NHA7,uniprot:A6NI34,uniprot:B2R9M0,uniprot:O00544,uniprot:Q9UQS1'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000002', 'input_common_names': '1,3-Diaminopropane', 'rxn_partner_common_name': 'DHPS', 'rxn_partner_ids': 'hmdb:HMDBP00219,gene_symbol:DHPS,uniprot:P49366,entrez:1725,ensembl:ENSG00000095059,uniprot:A0A024R7E9,uniprot:M0QZT3,uniprot:A0A024R7D0,uniprot:M0R264,uniprot:M0QYZ7,uniprot:Q5J8M5,uniprot:M0R0J4,uniprot:M0R1T4,uniprot:M0QX43,uniprot:M0R0V2,uniprot:M0R1T2,uniprot:M0QXT2,uniprot:B4E3M2,uniprot:M0R253,uniprot:A8K688,uniprot:M0R1I5,uniprot:Q13184,uniprot:Q13276,uniprot:Q9UDG0'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000002', 'input_common_names': '1,3-Diaminopropane', 'rxn_partner_common_name': 'ABP1,AOC1', 'rxn_partner_ids': 'hmdb:HMDBP00390,gene_symbol:ABP1,uniprot:P19801,gene_symbol:AOC1,entrez:26,ensembl:ENSG00000002726,uniprot:C9J2J4,uniprot:C9J0G8,uniprot:C9J690,uniprot:Q16683,uniprot:Q16684,uniprot:Q56II4,uniprot:Q6GU42'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000002', 'input_common_names': '1,3-Diaminopropane', 'rxn_partner_common_name': 'AOC3', 'rxn_partner_ids': 'hmdb:HMDBP00393,gene_symbol:AOC3,uniprot:Q16853,entrez:8639,ensembl:ENSG00000131471,uniprot:K7EL47,uniprot:K7EIK5,uniprot:K7EQZ5,uniprot:B2RCI5,uniprot:K7ESB3,uniprot:L0L8N9,uniprot:Q45F94'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000002', 'input_common_names': '1,3-Diaminopropane', 'rxn_partner_common_name': 'AOC2', 'rxn_partner_ids': 'hmdb:HMDBP00394,gene_symbol:AOC2,uniprot:O75106,entrez:314,ensembl:ENSG00000131480,uniprot:A5PKW2,uniprot:O00120,uniprot:O75105,uniprot:Q4TTW5,uniprot:Q9UNY0'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000002', 'input_common_names': '1,3-Diaminopropane', 'rxn_partner_common_name': 'AMD1', 'rxn_partner_ids': 'hmdb:HMDBP00420,gene_symbol:AMD1,uniprot:P17707,entrez:262,ensembl:ENSG00000123505,uniprot:Q5VXN5,uniprot:F6R5I9,uniprot:B4DZ60,uniprot:A0A5F9ZHD5,uniprot:E1P5F7,uniprot:Q5VXN4,uniprot:Q5VXN6,uniprot:Q9BWK4'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000002', 'input_common_names': '1,3-Diaminopropane', 'rxn_partner_common_name': 'ODC1', 'rxn_partner_ids': 'hmdb:HMDBP01034,gene_symbol:ODC1,uniprot:P11926,entrez:4953,ensembl:ENSG00000115758,uniprot:C9JG30,uniprot:Q53TU3,uniprot:Q6LDS9'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000001', 'input_common_names': '1-Methylhistidine', 'rxn_partner_common_name': 'CARNS1', 'rxn_partner_ids': 'hmdb:HMDBP08844,gene_symbol:CARNS1,uniprot:A5YM72,entrez:57571,ensembl:ENSG00000172508,uniprot:A8K1M3,uniprot:B4DFC6,uniprot:E9PK38,uniprot:F5H427,uniprot:Q8N467,uniprot:Q9P2F3'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000001', 'input_common_names': '1-Methylhistidine', 'rxn_partner_common_name': 'HEMK1', 'rxn_partner_ids': 'hmdb:HMDBP11778,gene_symbol:HEMK1,uniprot:Q9Y5R4'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000001', 'input_common_names': '1-Methylhistidine', 'rxn_partner_common_name': 'METTL2B', 'rxn_partner_ids': 'hmdb:HMDBP11824,gene_symbol:METTL2B,uniprot:Q6P1Q9,uniprot:B4DZ68,uniprot:Q0IJ54,uniprot:Q3B7J1'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000001', 'input_common_names': '1-Methylhistidine', 'rxn_partner_common_name': 'METTL6', 'rxn_partner_ids': 'hmdb:HMDBP11825,gene_symbol:METTL6,uniprot:Q8TCB7,uniprot:Q96LU4'}, {'query_relation': 'met2gene', 'input_analyte': 'hmdb:HMDB0000001', 'input_common_names': '1-Methylhistidine', 'rxn_partner_common_name': 'WBSCR22,BUD23', 'rxn_partner_ids': 'hmdb:HMDBP12007,gene_symbol:WBSCR22,uniprot:O43709,gene_symbol:BUD23,entrez:114049,ensembl:ENSG00000071462,uniprot:F8WD67,uniprot:B3KR83,uniprot:F2Z3D9,uniprot:H7C0G4,uniprot:F8W1Y1,uniprot:E9PCQ7,uniprot:H7C170,uniprot:F8WD87,uniprot:H7C439,uniprot:F8VZ18,uniprot:A8K501,uniprot:C9K060,uniprot:Q96P12,uniprot:Q9BQ58,uniprot:Q9HBP9'}, {'query_relation': 'gene2met', 'input_analyte': 'uniprot:P31323', 'input_common_names': 'PRKAR2B', 'rxn_partner_common_name': \"Cyclic AMP,IP-activated  cAMP,cAMP,3',5'-cyclic AMP(1-),3',5'-cyclic AMP\", 'rxn_partner_ids': 'hmdb:HMDB0000058,chebi:17489,chemspider:5851,kegg:C00575,pubchem:6076,CAS:60-92-4,hmdb:HMDB00058,wikidata:Q210041,chebi:58165,chemspider:5415746,pubchem:7059571,wikidata:Q27125416'}], 'function_call': ['RaMP::rampFastCata(c(\"hmdb:HMDB0000001\", \"hmdb:HMDB0000002\", \"uniprot:P31323\", \"uniprot:P04637\"))\\n'], 'numFoundIds': [3]}\n",
      "Failed to get reactions after 3 attempts: Request failed: 500 Server Error: Internal Server Error for url: https://rampdb.nih.gov/api/reactions-from-analytes\n",
      "Failed to get reaction classes after 3 attempts: Request failed: 500 Server Error: Internal Server Error for url: https://rampdb.nih.gov/api/reaction-classes-from-analytes\n",
      "\n",
      "=== Ontology Analysis ===\n",
      "\n",
      "Ontologies for metabolites:\n",
      "{'data': [{'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Saliva', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Feces', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Cytoplasm', 'HMDBOntologyType': 'Subcellular'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Urine', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Skeletal muscle', 'HMDBOntologyType': 'Tissue and substructures'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Blood', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Prostate', 'HMDBOntologyType': 'Organ and components'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Cerebrospinal fluid', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Pharmaceutical', 'HMDBOntologyType': 'Industrial application'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Endogenous', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Diabetes mellitus type 2', 'HMDBOntologyType': 'Health condition'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Muscle', 'HMDBOntologyType': 'Tissue and substructures'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Obesity', 'HMDBOntologyType': 'Health condition'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Placenta', 'HMDBOntologyType': 'Tissue and substructures'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Pharmaceutical industry', 'HMDBOntologyType': 'Industrial application'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Kidney disease', 'HMDBOntologyType': 'Health condition'}, {'Metabolites': '1-Methylhistidine', 'sourceId': 'hmdb:HMDB0000001', 'IDtype': 'hmdb', 'Ontology': 'Metabolism and nutrition disorders', 'HMDBOntologyType': 'Health condition'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Feces', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Feces', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Urine', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Urine', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Blood', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Blood', 'HMDBOntologyType': 'Biofluid and excreta'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Endogenous', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Endogenous', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Cytoplasm', 'HMDBOntologyType': 'Subcellular'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Cytoplasm', 'HMDBOntologyType': 'Subcellular'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Food', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Food', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Plant', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Plant', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Animal', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Animal', 'HMDBOntologyType': 'Source'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Cancer', 'HMDBOntologyType': 'Health condition'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Cancer', 'HMDBOntologyType': 'Health condition'}, {'Metabolites': '1,3-Diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Leukemia', 'HMDBOntologyType': 'Health condition'}, {'Metabolites': '1,3-diaminopropane', 'sourceId': 'hmdb:HMDB0000002', 'IDtype': 'hmdb', 'Ontology': 'Leukemia', 'HMDBOntologyType': 'Health condition'}], 'function_call': ['RaMP::getOntoFromMeta(c(\"hmdb:HMDB0000001\", \"hmdb:HMDB0000002\"))\\n'], 'numFoundIds': [2]}\n",
      "Failed to get metabolites from ontologies after 3 attempts: Request failed: 500 Server Error: Internal Server Error for url: https://rampdb.nih.gov/api/metabolites-from-ontologies\n",
      "\n",
      "=== Enrichment Analysis ===\n",
      "\n",
      "Chemical enrichment results:\n",
      "{'data': {'ClassyFire_class': [{'category': 'ClassyFire_class', 'class_name': 'Organonitrogen compounds', 'met_hits': 1, 'pop_hits': 685, 'met_size': 2, 'pop_size': 145833, 'p-value': 0.0094, 'adjP_BH': 0.0187}, {'category': 'ClassyFire_class', 'class_name': 'Carboxylic acids and derivatives', 'met_hits': 1, 'pop_hits': 4270, 'met_size': 2, 'pop_size': 145833, 'p-value': 0.0577, 'adjP_BH': 0.0577}], 'ClassyFire_sub_class': [{'category': 'ClassyFire_sub_class', 'class_name': 'Amines', 'met_hits': 1, 'pop_hits': 375, 'met_size': 2, 'pop_size': 140755, 'p-value': 0.0053, 'adjP_BH': 0.0106}, {'category': 'ClassyFire_sub_class', 'class_name': 'Amino acids, peptides, and analogues', 'met_hits': 1, 'pop_hits': 3611, 'met_size': 2, 'pop_size': 140755, 'p-value': 0.0507, 'adjP_BH': 0.0507}], 'ClassyFire_super_class': [{'category': 'ClassyFire_super_class', 'class_name': 'Organic nitrogen compounds', 'met_hits': 1, 'pop_hits': 685, 'met_size': 2, 'pop_size': 145937, 'p-value': 0.0094, 'adjP_BH': 0.0187}, {'category': 'ClassyFire_super_class', 'class_name': 'Organic acids and derivatives', 'met_hits': 1, 'pop_hits': 5311, 'met_size': 2, 'pop_size': 145937, 'p-value': 0.0715, 'adjP_BH': 0.0715}], 'result_type': ['chemical_class_enrichment']}}\n"
     ]
    }
   ],
   "source": [
    "# RaMP-DB API Client Tutorial\n",
    "\n",
    "\"\"\"\n",
    "This tutorial demonstrates how to use the RaMP-DB API client module\n",
    "for metabolic pathway and biochemical data analysis.\n",
    "\"\"\"\n",
    "\n",
    "from typing import List, Dict\n",
    "from ramp_client import RaMPClient, RaMPConfig, RaMPAPIError\n",
    "import time\n",
    "\n",
    "def main():\n",
    "    # 1. Initialize the client\n",
    "    # You can use default settings\n",
    "    client = RaMPClient()\n",
    "    \n",
    "    # Or customize the configuration\n",
    "    custom_config = RaMPConfig(\n",
    "        base_url=\"https://rampdb.nih.gov/api\",\n",
    "        timeout=60  # Extended timeout for larger queries\n",
    "    )\n",
    "    client_custom = RaMPClient(config=custom_config)\n",
    "\n",
    "    # 2. Basic Database Information\n",
    "    try:\n",
    "        # Get source database versions\n",
    "        versions = client.get_source_versions()\n",
    "        print(\"Database versions:\", versions)\n",
    "\n",
    "        # Get valid ID types\n",
    "        id_types = client.get_id_types()\n",
    "        print(\"Valid ID types:\", id_types)\n",
    "    except RaMPAPIError as e:\n",
    "        print(f\"Error getting database info: {e}\")\n",
    "\n",
    "    # 3. Metabolite Analysis Example\n",
    "    def analyze_metabolites(metabolite_ids: List[str]):\n",
    "        try:\n",
    "            # Add 'hmdb:' prefix if not present\n",
    "            formatted_ids = [\n",
    "                f\"hmdb:{mid}\" if not mid.startswith(\"hmdb:\") else mid \n",
    "                for mid in metabolite_ids\n",
    "            ]\n",
    "            \n",
    "            # Get chemical properties\n",
    "            properties = client.get_chemical_properties(formatted_ids)\n",
    "            print(\"\\nChemical properties:\")\n",
    "            print(properties)\n",
    "\n",
    "            # Get chemical classes\n",
    "            classes = client.get_chemical_classes(formatted_ids)\n",
    "            print(\"\\nChemical classes:\")\n",
    "            print(classes)\n",
    "\n",
    "            # Get pathways\n",
    "            pathways = client.get_pathways_from_analytes(formatted_ids)\n",
    "            \n",
    "            # Analyze pathway statistics\n",
    "            if \"result\" in pathways and pathways[\"result\"]:\n",
    "                stats = client.analyze_pathway_stats(pathways)\n",
    "                for metabolite_id, pathway_stats in stats.items():\n",
    "                    print(f\"\\nPathway stats for {metabolite_id}:\")\n",
    "                    print(f\"Total pathways: {pathway_stats.total_pathways}\")\n",
    "                    print(f\"Pathways by source: {pathway_stats.pathways_by_source}\")\n",
    "                    print(f\"Unique pathway names: {len(pathway_stats.unique_pathway_names)}\")\n",
    "                    print(f\"Pathway sources: {pathway_stats.pathway_sources}\")\n",
    "\n",
    "                # Find pathway overlaps\n",
    "                overlaps = client.find_pathway_overlaps(pathways)\n",
    "                print(\"\\nPathway overlaps:\")\n",
    "                for pathway, count in overlaps.items():\n",
    "                    if count > 1:  # Show only pathways shared by multiple metabolites\n",
    "                        print(f\"{pathway}: shared by {count} metabolites\")\n",
    "            else:\n",
    "                print(\"No pathway data found\")\n",
    "\n",
    "        except RaMPAPIError as e:\n",
    "            print(f\"Error in metabolite analysis: {e}\")\n",
    "\n",
    "    # 4. Protein Pathway Analysis Example\n",
    "    def analyze_protein_pathways(protein_ids: List[str]):\n",
    "        try:\n",
    "            # Format protein IDs with uniprot prefix for pathway analysis\n",
    "            formatted_protein_ids = [\n",
    "                f\"uniprot:{pid}\" if not pid.startswith(\"uniprot:\") else pid \n",
    "                for pid in protein_ids\n",
    "            ]\n",
    "            \n",
    "            # Get pathways associated with proteins\n",
    "            protein_pathways = client.get_pathways_from_analytes(formatted_protein_ids)\n",
    "            print(\"\\nProtein pathways:\")\n",
    "            print(protein_pathways)\n",
    "\n",
    "            # Analyze pathway statistics\n",
    "            if \"result\" in protein_pathways and protein_pathways[\"result\"]:\n",
    "                stats = client.analyze_pathway_stats(protein_pathways)\n",
    "                for protein_id, pathway_stats in stats.items():\n",
    "                    print(f\"\\nPathway stats for {protein_id}:\")\n",
    "                    print(f\"Total pathways: {pathway_stats.total_pathways}\")\n",
    "                    print(f\"Pathways by source: {pathway_stats.pathways_by_source}\")\n",
    "                    print(f\"Unique pathway names: {len(pathway_stats.unique_pathway_names)}\")\n",
    "                    print(f\"Pathway sources: {pathway_stats.pathway_sources}\")\n",
    "            else:\n",
    "                print(\"No pathway data found\")\n",
    "\n",
    "        except RaMPAPIError as e:\n",
    "            print(f\"Error in protein pathway analysis: {e}\")\n",
    "\n",
    "    # 5. Combined Analysis Example\n",
    "    def analyze_metabolites_and_proteins(\n",
    "        metabolite_ids: List[str], protein_ids: List[str]\n",
    "    ):\n",
    "        try:\n",
    "            # Combine IDs with appropriate prefixes\n",
    "            combined_analytes = (\n",
    "                [f\"hmdb:{mid}\" if not mid.startswith(\"hmdb:\") else mid \n",
    "                 for mid in metabolite_ids] +\n",
    "                [f\"uniprot:{pid}\" if not pid.startswith(\"uniprot:\") else pid \n",
    "                 for pid in protein_ids]\n",
    "            )\n",
    "\n",
    "            # Get common reaction analytes\n",
    "            common_reactions = client.get_common_reaction_analytes(combined_analytes)\n",
    "            print(\"\\nCommon reaction analytes:\")\n",
    "            print(common_reactions)\n",
    "\n",
    "            # Attempt to get reactions with retry\n",
    "            max_retries = 3\n",
    "            for attempt in range(max_retries):\n",
    "                try:\n",
    "                    reactions = client.get_reactions_from_analytes(combined_analytes)\n",
    "                    print(\"\\nReactions:\")\n",
    "                    print(reactions)\n",
    "                    break\n",
    "                except RaMPAPIError as e:\n",
    "                    if attempt == max_retries - 1:\n",
    "                        print(f\"Failed to get reactions after {max_retries} attempts: {e}\")\n",
    "                    else:\n",
    "                        time.sleep(2)  # Wait before retry\n",
    "\n",
    "            # Attempt to get reaction classes with retry\n",
    "            for attempt in range(max_retries):\n",
    "                try:\n",
    "                    reaction_classes = client.get_reaction_classes(combined_analytes)\n",
    "                    print(\"\\nReaction classes:\")\n",
    "                    print(reaction_classes)\n",
    "                    break\n",
    "                except RaMPAPIError as e:\n",
    "                    if attempt == max_retries - 1:\n",
    "                        print(f\"Failed to get reaction classes after {max_retries} attempts: {e}\")\n",
    "                    else:\n",
    "                        time.sleep(2)  # Wait before retry\n",
    "\n",
    "        except RaMPAPIError as e:\n",
    "            print(f\"Error in combined analysis: {e}\")\n",
    "\n",
    "    # 6. Ontology Analysis Example\n",
    "    def analyze_ontologies(metabolite_ids: List[str], ontology_terms: List[str]):\n",
    "        try:\n",
    "            # Format metabolite IDs\n",
    "            formatted_ids = [\n",
    "                f\"hmdb:{mid}\" if not mid.startswith(\"hmdb:\") else mid \n",
    "                for mid in metabolite_ids\n",
    "            ]\n",
    "            \n",
    "            # Get ontologies from metabolites\n",
    "            ontologies = client.get_ontologies_from_metabolites(\n",
    "                formatted_ids, names_or_ids=\"ids\"\n",
    "            )\n",
    "            print(\"\\nOntologies for metabolites:\")\n",
    "            print(ontologies)\n",
    "\n",
    "            # Attempt to get metabolites from ontologies with retry\n",
    "            max_retries = 3\n",
    "            for attempt in range(max_retries):\n",
    "                try:\n",
    "                    metabolites = client.get_metabolites_from_ontologies(\n",
    "                        ontology_terms, output_format=\"json\"\n",
    "                    )\n",
    "                    print(\"\\nMetabolites for ontologies:\")\n",
    "                    print(metabolites)\n",
    "                    break\n",
    "                except RaMPAPIError as e:\n",
    "                    if attempt == max_retries - 1:\n",
    "                        print(f\"Failed to get metabolites from ontologies after {max_retries} attempts: {e}\")\n",
    "                    else:\n",
    "                        time.sleep(2)  # Wait before retry\n",
    "\n",
    "        except RaMPAPIError as e:\n",
    "            print(f\"Error in ontology analysis: {e}\")\n",
    "\n",
    "    # 7. Chemical Enrichment Analysis Example\n",
    "    def perform_enrichment_analysis(metabolite_ids: List[str]):\n",
    "        try:\n",
    "            # Format metabolite IDs\n",
    "            formatted_ids = [\n",
    "                f\"hmdb:{mid}\" if not mid.startswith(\"hmdb:\") else mid \n",
    "                for mid in metabolite_ids\n",
    "            ]\n",
    "            \n",
    "            # Attempt enrichment analysis with retry\n",
    "            max_retries = 3\n",
    "            for attempt in range(max_retries):\n",
    "                try:\n",
    "                    enrichment_results = client.perform_chemical_enrichment(formatted_ids)\n",
    "                    print(\"\\nChemical enrichment results:\")\n",
    "                    print(enrichment_results)\n",
    "                    break\n",
    "                except RaMPAPIError as e:\n",
    "                    if attempt == max_retries - 1:\n",
    "                        print(f\"Failed to perform enrichment analysis after {max_retries} attempts: {e}\")\n",
    "                    else:\n",
    "                        time.sleep(2)  # Wait before retry\n",
    "\n",
    "        except RaMPAPIError as e:\n",
    "            print(f\"Error in enrichment analysis: {e}\")\n",
    "\n",
    "    # Example usage with real metabolite and protein IDs\n",
    "    metabolite_ids = [\"HMDB0000001\", \"HMDB0000002\"]  # Example metabolite IDs\n",
    "    protein_ids = [\"P31323\", \"P04637\"]  # Example protein IDs\n",
    "    ontology_terms = [\"CHEBI:15903\"]  # Example ontology terms\n",
    "\n",
    "    print(\"\\n=== Metabolite Analysis ===\")\n",
    "    analyze_metabolites(metabolite_ids)\n",
    "\n",
    "    print(\"\\n=== Protein Pathway Analysis ===\")\n",
    "    analyze_protein_pathways(protein_ids)\n",
    "\n",
    "    print(\"\\n=== Combined Analysis ===\")\n",
    "    analyze_metabolites_and_proteins(metabolite_ids, protein_ids)\n",
    "\n",
    "    print(\"\\n=== Ontology Analysis ===\")\n",
    "    analyze_ontologies(metabolite_ids, ontology_terms)\n",
    "\n",
    "    print(\"\\n=== Enrichment Analysis ===\")\n",
    "    perform_enrichment_analysis(metabolite_ids)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
