{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "54013cf5",
   "metadata": {},
   "source": [
    "# Generate masks, check masks\n",
    "___\n",
    "### 1. Organize images into npy stacks.\n",
    "### 2. Preprocess images for object detection.\n",
    "### 3. Generate masks.\n",
    "### 4. Check masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7f4ba39",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.plot import preprocess_generate_masks\n",
    "%matplotlib inline\n",
    "\n",
    "settings = {'metadata_type':'cellvoyager',\n",
    "            'custom_regex':None,\n",
    "            'experiment':'ubiquitin',\n",
    "            'channels':[0,1,2,3],\n",
    "            'nucleus_channel':0,\n",
    "            'nucleus_background':500,\n",
    "            'nucleus_Signal_to_noise':2,\n",
    "            'nucleus_CP_prob':0,\n",
    "            'remove_background_nucleus':False,\n",
    "            'cell_channel':3,\n",
    "            'cell_background':100,\n",
    "            'cell_Signal_to_noise':10,\n",
    "            'cell_CP_prob':-1,\n",
    "            'remove_background_cell':False,\n",
    "            'pathogen_model':None,\n",
    "            'pathogen_channel':1,\n",
    "            'pathogen_background':200,\n",
    "            'pathogen_Signal_to_noise':5,\n",
    "            'pathogen_CP_prob':-1,\n",
    "            'remove_background_pathogen':True,\n",
    "            'magnification':40,\n",
    "            'save':True,\n",
    "            'preprocess':True,\n",
    "            'masks':True,\n",
    "            'batch_size':50,\n",
    "            'filter':False,\n",
    "            'merge_pathogens':True,\n",
    "            'plot':False,\n",
    "            'adjust_cells':True,\n",
    "            'test_mode':False,\n",
    "            'test_images':10,\n",
    "            'random_test':True}\n",
    "\n",
    "src = 'path to folder with images'\n",
    "spacr.core.preprocess_generate_masks(src, settings=settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40426788",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.plot import plot_images_and_arrays\n",
    "%matplotlib inline\n",
    "\n",
    "folder_paths = ['path to folder that origionally contained images', ...]\n",
    "\n",
    "plot_images_and_arrays(folder_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08cd3d1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.plot import plot_merged\n",
    "%matplotlib inline\n",
    "\n",
    "settings = {'include_noninfected':True, \n",
    "            'include_multiinfected':True,\n",
    "            'include_multinucleated':True,\n",
    "            'remove_background':False,\n",
    "            'filter_min_max':[[0,1000000],[0,1000000],[0,1000000]],\n",
    "            'channel_dims':[0,1,2,3],\n",
    "            'backgrounds':[100,100,100,100],\n",
    "            'cell_mask_dim':4,\n",
    "            'nucleus_mask_dim':5,\n",
    "            'pathogen_mask_dim':6,\n",
    "            'overlay_chans':[0,2,3],\n",
    "            'outline_thickness':3,\n",
    "            'outline_color':'gbr',\n",
    "            'overlay':True,\n",
    "            'normalization_percentiles':[2,98],\n",
    "            'normalize':False,\n",
    "            'print_object_number':True,\n",
    "            'nr':1,\n",
    "            'figuresize':20,\n",
    "            'cmap':'inferno',\n",
    "            'verbose':True}\n",
    "\n",
    "src = 'path to folder with merged images and masks'\n",
    "fig = plot_merged(src, settings)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc3d5811",
   "metadata": {},
   "source": [
    "# Generate Measurements, single cell images,  ML scores\n",
    "___\n",
    "### 1. Collect measurement data and save to database, Generate single object images.\n",
    "### 2. Generate ML scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a699d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "settings = {'input_folder':'path to folder with merged images and masks',\n",
    "            'channels':[0,1,2,3],\n",
    "            'cell_mask_dim':4,\n",
    "            'cell_min_size':2000,\n",
    "            'nucleus_mask_dim':5,\n",
    "            'nucleus_min_size':1000,\n",
    "            'pathogen_mask_dim':6,\n",
    "            'pathogen_min_size':400,\n",
    "            'cytoplasm_min_size':0,\n",
    "            'save_png':True,\n",
    "            'crop_mode':['cell'],\n",
    "            'use_bounding_box':False,\n",
    "            'png_size':[[224,224]],\n",
    "            'normalize':False,\n",
    "            'png_dims':[0,1,2],\n",
    "            'normalize_by':'fov', #'fov' or 'png'\n",
    "            'save_measurements':True,\n",
    "            'plot':False,\n",
    "            'plot_filtration':False,\n",
    "            'include_uninfected':False,\n",
    "            'test_mode':False}\n",
    "\n",
    "spacr.measure.measure_crop(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23c16e32",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import generate_ml_scores\n",
    "\n",
    "settings = {'model_type':'xgboost',\n",
    "            'heatmap_feature':'predictions',\n",
    "            'grouping':'mean',\n",
    "            'min_max':'allq',\n",
    "            'cmap':'viridis',\n",
    "            'channel_of_interest':3,\n",
    "            'minimum_cell_count':25,\n",
    "            'n_estimators':100,\n",
    "            'test_size':0.2,\n",
    "            'location_column':'col',\n",
    "            'positive_control':'c2',\n",
    "            'negative_control':'c1',\n",
    "            'exclude':None,\n",
    "            'n_repeats':10,\n",
    "            'top_features':30,\n",
    "            'remove_low_variance_features':True,\n",
    "            'remove_highly_correlated_features':True,\n",
    "            'n_jobs':-1,\n",
    "            'verbose':True}\n",
    "\n",
    "src = 'path to folder that origionally contained images'\n",
    "results = generate_ml_scores(src, settings)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a2daa6c",
   "metadata": {},
   "source": [
    "# Auxiliary functions\n",
    "___\n",
    "### 1. annotate_app_v2: Annotate images after exporting measurements and saving PNGs. annotations are saved to the measurements.db in the table png_list\n",
    "### 2. analyze_recruitment: Read the tables in measurements.db to a pd.DataFrame, annotate conditions and evaluate recruitment.\n",
    "### 3. analyze_plaques: Generate a csv file with the area and count of objects in masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a348105f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.app_annotate import annotate\n",
    "\n",
    "settings = {'src':'path to folder that origionally contained images',\n",
    "            'image_type':'cell_png',\n",
    "            'channels':['r', 'g', 'b'],\n",
    "            'geom': \"3200x2000\",\n",
    "            'img_size':(200, 200),\n",
    "            'rows':10,\n",
    "            'columns':18,\n",
    "            'annotation_column':'recruited_test',\n",
    "            'normalize':False,\n",
    "            'percentiles':(2,98),\n",
    "            'measurement':None,\n",
    "            'threshold':None}\n",
    "\n",
    "annotate(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbecfd53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate recruitment.\n",
    "from spacr.core import analyze_recruitment\n",
    "%matplotlib inline\n",
    "\n",
    "metadata_settings = {'target':'protein',\n",
    "                     'cell_types':['HeLa'],\n",
    "                     'cell_plate_metadata':None,\n",
    "                     'pathogen_types':['pathogen_1', 'pathogen_2'],\n",
    "                     'pathogen_plate_metadata':[['c1', 'c2', 'c3'],['c4','c5', 'c6']],\n",
    "                     'treatments':['cm', 'lovastatin'],\n",
    "                     'treatment_plate_metadata':[['r1', 'r2','r3'], ['r4', 'r5','r6']],\n",
    "                     'metadata_types':['col', 'col', 'row'],\n",
    "                     'channel_dims':[0,1,2,3],\n",
    "                     'cell_chann_dim':3,\n",
    "                     'cell_mask_dim':4,\n",
    "                     'nucleus_chann_dim':0,\n",
    "                     'nucleus_mask_dim':5,\n",
    "                     'pathogen_chann_dim':2,\n",
    "                     'pathogen_mask_dim':6,\n",
    "                     'channel_of_interest':2,\n",
    "                     'plot':True,\n",
    "                     'plot_nr':1,\n",
    "                     'plot_control':True,\n",
    "                     'figuresize':20,\n",
    "                     'remove_background':True,\n",
    "                     'backgrounds':100,\n",
    "                     'include_noninfected':False,\n",
    "                     'include_multiinfected':2,\n",
    "                     'include_multinucleated':False,\n",
    "                     'cells_per_well':100,\n",
    "                     'pathogen_size_range':[0,100000],\n",
    "                     'nucleus_size_range':[0,100000],\n",
    "                     'cell_size_range':[0,100000],\n",
    "                     'pathogen_intensity_range':[0,100000],\n",
    "                     'nucleus_intensity_range':[0,100000],\n",
    "                     'cell_intensity_range':[0,100000],\n",
    "                     'target_intensity_min':0}\n",
    "\n",
    "src = 'path to folder that origionally contained images'\n",
    "\n",
    "dfs = analyze_recruitment(src, settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac02c974",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import analyze_plaques\n",
    "\n",
    "folder = 'path to folder with plaque masks'\n",
    "\n",
    "analyze_plaques(folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76f877e8",
   "metadata": {},
   "source": [
    "# Train computer vision Torch models\n",
    "___\n",
    "### 1. generate_training_dataset: Generate train and test folders with class subfolders for training DL models.\n",
    "### 2. train_test_model: Train torch model\n",
    "### 3. generate_dataset: Generate dataset to apply model to\n",
    "### 4. apply_model_to_tar: Apply trained model to tar dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0b50d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import generate_training_dataset\n",
    "\n",
    "src = 'path to folder that origionally contained images'\n",
    "generate_training_dataset(src,\n",
    "                          mode='metadata', # annotation, recruitment or metadata\n",
    "                          annotation_column='test', # for mode = annotation\n",
    "                          annotated_classes=[1,2], # for mode = annotation\n",
    "                          classes=['nc','pc'], # for mode = annotation\n",
    "                          size=None, # number of images per class; None = nr of images in least abundant class \n",
    "                          test_split=0.1,\n",
    "                          metadata_type_by='col',\n",
    "                          class_metadata=[['c1'],['c2']], # for mode = metadata\n",
    "                          channel_of_interest=3) # for mode = recruitment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4b32a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.deep_spacr import train_test_model\n",
    "%matplotlib inline\n",
    "\n",
    "settings = {'train':True,\n",
    "            'test': False,\n",
    "            'classes':['nc','pc'],\n",
    "            'model_type':'maxvit_t',\n",
    "            'optimizer_type':'adamw',\n",
    "            'schedule':'reduce_lr_on_plateau', # reduce_lr_on_plateau, step_lr\n",
    "            'loss_type':'focal_loss', # binary_cross_entropy_with_logits, focal_loss\n",
    "            'normalize':True,\n",
    "            'image_size':224,\n",
    "            'batch_size':64,\n",
    "            'epochs':100,\n",
    "            'val_split':0.1,\n",
    "            'train_mode':'erm',\n",
    "            'learning_rate':0.0001,\n",
    "            'weight_decay':0.00001,\n",
    "            'dropout_rate':0.1,\n",
    "            'init_weights':True,\n",
    "            'amsgrad':True,\n",
    "            'use_checkpoint':True,\n",
    "            'gradient_accumulation':True,\n",
    "            'gradient_accumulation_steps':4,\n",
    "            'intermedeate_save':True,\n",
    "            'pin_memory':True,\n",
    "            'num_workers':30,\n",
    "            'channels':['r','g','b'],\n",
    "            'verbose':True}\n",
    "\n",
    "src = 'path to training dataset folder, basename training'\n",
    "train_test_model(src, settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc8885a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import generate_dataset\n",
    "\n",
    "src = 'path to folder that origionally contained images'\n",
    "generate_dataset(src,\n",
    "                 file_metadata=None,\n",
    "                 experiment='experiment_1',\n",
    "                 sample=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccc52a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import apply_model_to_tar\n",
    "%matplotlib inline\n",
    "\n",
    "tar_path = 'path to dataset, extention is .tar'\n",
    "model_path='path to model, extention is .pth'\n",
    "\n",
    "result_df = apply_model_to_tar(tar_path,\n",
    "                               model_path,\n",
    "                               file_type='cell_png',\n",
    "                               image_size=224,\n",
    "                               batch_size=64,\n",
    "                               normalize=True,\n",
    "                               preload='images',\n",
    "                               num_workers=30,\n",
    "                               threshold=0.5,\n",
    "                               verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18221009",
   "metadata": {},
   "source": [
    "# Analyze sequencing reeds\n",
    "___\n",
    "### 1. analyze_reads: Merge paired end reads and extract barecodes.\n",
    "### 2. map_barcodes_folder: Map barecode and gRNA sequences to names\n",
    "### 3. plot_lorenz_curves: Generate lorenz curves from unique_grna.csv\n",
    "### 4. perform_regression: Perform multiple linear regression to estimate the effect size of gRNAs on the dependent variable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d58f97a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.sequencing import analyze_reads\n",
    "\n",
    "settings = {'src':'path to folder with FASTQ files',\n",
    "           'upstream':'CTTCTGGTAAATGGGGATGTCAAGTT',\n",
    "           'downstream':'GTTTAAGAGCTATGCTGGAAACAGCAG',\n",
    "           'barecode_length_1':8,\n",
    "           'barecode_length_2':7,\n",
    "           'chunk_size':1000000,\n",
    "           'test':False}\n",
    "\n",
    "analyze_reads(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef2eeb3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.sequencing import map_barcodes_folder\n",
    "\n",
    "src = 'path to folder containing .h5 files, .*_combined.h5'\n",
    "\n",
    "plate_name_dict = {'E01': 'plate1','E02': 'plate2'}\n",
    "\n",
    "settings = {'pc':'positive control gRNA name',\n",
    "            'pc_loc':'positive control column',\n",
    "            'nc':'negative control gRNA name',\n",
    "            'nc_loc':'negative control column'}\n",
    "            'grna', 'path to grna sequence | name csv file')\n",
    "            'barcodes', 'path to barecode sequence | name csv file')\n",
    "            'plate_dict', plate_name_dict)\n",
    "            'test', False)\n",
    "            'verbose', True}\n",
    "            \n",
    "map_barcodes_folder(src, settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "785c7519",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.plot import plot_lorenz_curves\n",
    "%matplotlib inline\n",
    "\n",
    "csv_files = ['path to csv file (.*_combined_unique_grna.csv)', 'path to csv file (.*_combined_unique_grna.csv)']\n",
    "plot_lorenz_curves(csv_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31af577f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.sequencing import perform_regression\n",
    "%matplotlib inline\n",
    "\n",
    "csv_path = 'path to reads, .*combined_combination_counts.csv'\n",
    "dependent_variable = 'path do computer vision results'\n",
    "\n",
    "settings = {'dependent_variable':dependent_variable,\n",
    "            'gene_weights_csv':csv_path,\n",
    "            'fraction_threshold':0.1,\n",
    "            'dependent_variable':'prediction_probability_class_1',\n",
    "            'transform':None,\n",
    "            'agg_type':'mean',\n",
    "            'min_cell_count':25,\n",
    "            'regression_type':'ols',\n",
    "            'remove_row_column_effect':False,\n",
    "            'alpha':1,\n",
    "            'nc':'c1',\n",
    "            'pc':'c2',\n",
    "            'other':'c3'}\n",
    "\n",
    "coef_df = perform_regression(settings)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "484a08cd",
   "metadata": {},
   "source": [
    "# Simulate Pooled screens\n",
    "___\n",
    "### 1. run_and_save: Simulate pooled screen\n",
    "### 2. run_multiple_simulations: Simulate multiple screens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cbaff87",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.sim import run_and_save\n",
    "\n",
    "#For 1 simulation\n",
    "settings = {\n",
    "    'name':'test',\n",
    "    'variable':'none',\n",
    "    'src': 'path',\n",
    "    'plot': True,\n",
    "    'random_seed': True,\n",
    "    'nr_plates': 4,\n",
    "    'number_of_genes': 100,\n",
    "    'number_of_active_genes': 10,\n",
    "    'number_of_control_genes': 10,\n",
    "    'avg_genes_per_well': 5,\n",
    "    'sd_genes_per_well':2,\n",
    "    'avg_cells_per_well': 100,\n",
    "    'sd_cells_per_well':50,\n",
    "    'positive_mean': 0.8,\n",
    "    'positive_variance':0.15,\n",
    "    'negative_mean':0.2,\n",
    "    'negative_variance':0.15,\n",
    "    'avg_reads_per_gene': 1000,\n",
    "    'sd_reads_per_gene':500,\n",
    "    'sequencing_error': 0.1,\n",
    "    'well_ineq_coeff': 0.5,\n",
    "    'gene_ineq_coeff': 0.1,\n",
    "    'max_workers': 25}\n",
    "\n",
    "i=1\n",
    "run_and_save(i, settings, time_ls=[1], total_sims=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73a3d903",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.sim import run_multiple_simulations\n",
    "\n",
    "#For multiple simulation\n",
    "\n",
    "#lopp over most variables\n",
    "avg_genes_per_well_ls = [10]\n",
    "avg_cells_per_well_ls = [100]\n",
    "positive_mean_ls = [0.8]\n",
    "avg_reads_per_gene_ls = [1000]\n",
    "sequencing_error_ls = [0.01]\n",
    "well_ineq_coeff_ls = [0.3] #lower mean more inequality (0-1)\n",
    "gene_ineq_coeff_ls = [0.8]\n",
    "nr_plates_ls = [8]\n",
    "number_of_genes_ls = [1384]\n",
    "number_of_active_genes_ls = [8]\n",
    "\n",
    "settings = {\n",
    "    'name':'plates_2_4_8',\n",
    "    'variable':'all',\n",
    "    'src': 'path',\n",
    "    'plot': True,\n",
    "    'random_seed': False,\n",
    "    'replicates':2,\n",
    "    'nr_plates': nr_plates_ls,\n",
    "    'number_of_genes': number_of_genes_ls,\n",
    "    'number_of_active_genes': number_of_active_genes_ls,\n",
    "    'number_of_control_genes': 30,\n",
    "    'avg_genes_per_well': avg_genes_per_well_ls,\n",
    "    'avg_cells_per_well': avg_cells_per_well_ls,\n",
    "    'positive_mean': positive_mean_ls,\n",
    "    'avg_reads_per_gene': avg_reads_per_gene_ls,\n",
    "    'sequencing_error': sequencing_error_ls,\n",
    "    'well_ineq_coeff': well_ineq_coeff_ls,\n",
    "    'gene_ineq_coeff': gene_ineq_coeff_ls,\n",
    "    'max_workers': 25}\n",
    "\n",
    "run_multiple_simulations(settings)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5623759f",
   "metadata": {},
   "source": [
    "# Train Cellpose Models\n",
    "___\n",
    "### 1. sample_and_copy_images: Randomly select images from folders to generate an image dataset.\n",
    "### 2. generate_cellpose_train_test: Split images from a folder into test and train datasets.\n",
    "### 3. train_cellpose: Train cellpose models.\n",
    "### 4. check_cellpose_models: generate masks using all stock cellpose models.\n",
    "### 5. identify_masks_finetune: Use a custom or stock cellpose model to generate masks.\n",
    "### 6. compare_cellpose_masks: Compare masks generated with different cellpose models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95b77a44",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.alpha import sample_and_copy_images\n",
    "\n",
    "folder_list = ['path to golder with single channel images', ...]\n",
    "\n",
    "nr_of_images = 100\n",
    "dst = 'path to destination folder where images will be copied to'\n",
    "sample_and_copy_images(folder_list, nr_of_images, dst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "797805cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.io import generate_cellpose_train_test\n",
    "src = 'path to folder containing images that will be used for train and test datasets'\n",
    "generate_cellpose_train_test(src, test_split=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80a10287",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import train_cellpose\n",
    "%matplotlib inline\n",
    "\n",
    "settings = {'img_src':'path to train folder, basename train',\n",
    "            'model_name':'model_name',\n",
    "            'model_type':'cyto',\n",
    "            'Signal_to_noise':10,\n",
    "            'background':200,\n",
    "            'remove_background':False,\n",
    "            'learning_rate':0.2,\n",
    "            'weight_decay':1e-05,\n",
    "            'batch_size':8,\n",
    "            'n_epochs':25000,\n",
    "            'from_scratch':False,\n",
    "            'diameter':30,\n",
    "            'resize':True,\n",
    "            'width_height':[1120,1120],\n",
    "            'verbose':True}\n",
    "\n",
    "train_cellpose(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78c50677",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import check_cellpose_models\n",
    "%matplotlib inline\n",
    "\n",
    "settings = {'src':'path to test folder, basename test',\n",
    "           'batch_size':8,\n",
    "           'CP_prob':0,\n",
    "           'diameter':30,\n",
    "           'save':True,\n",
    "           'normalize':True,\n",
    "           'remove_background':False,\n",
    "           'background':200,\n",
    "           'Signal_to_noise':10,\n",
    "           'plot':True,\n",
    "           'verbose':True}\n",
    "\n",
    "check_cellpose_models(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dc14202",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import identify_masks_finetune\n",
    "%matplotlib inline\n",
    "\n",
    "settings = {'src':'path to folder with images',\n",
    "            'dst':'path to folder masks will be saved in',\n",
    "            'model_name':'cyto',\n",
    "            'custom_model': 'path to model, extention: .CP_model',\n",
    "            'background':200,\n",
    "            'Signal_to_noise':10,\n",
    "            'save':True,\n",
    "            'resize':True,\n",
    "            'target_height':1120,\n",
    "            'target_width':1120,\n",
    "            'diameter':30,\n",
    "            'CP_prob':0,\n",
    "            'invert':False,\n",
    "            'verbose':True}\n",
    "\n",
    "identify_masks_finetune(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a354a68",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import compare_cellpose_masks\n",
    "%matplotlib inline\n",
    "\n",
    "src = 'path to folder with images'\n",
    "compare_cellpose_masks(src, verbose=True, save=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c0032a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Trigger gui application\n",
    "___\n",
    "### 1. gui_make_masks: Modify or draw masks using single channel images.\n",
    "### 2. \n",
    "### 3. \n",
    "### 4. \n",
    "### 5. \n",
    "### 6. \n",
    "### 7. \n",
    "### 8. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a957e402",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.mask_app import gui_make_masks\n",
    "gui_make_masks()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ae71432",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.app_mask import start_mask_app\n",
    "start_mask_app()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbb387ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.app_measure import start_measure_app\n",
    "start_measure_app()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12630aba",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.app_classify import start_classify_app\n",
    "start_classify_app()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "961a2137-86bd-4276-a5a6-a312ed962270",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.app_sequencing import start_seq_app\n",
    "start_seq_app()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc4808d-a2ba-470a-bf91-9c2aa2a9449b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.app_umap import start_umap_app\n",
    "start_umap_app()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66a6715b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.gui import gui_app\n",
    "gui_app()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spacr",
   "language": "python",
   "name": "spacr"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
